api/envoy/type/substitution.proto

```diff
@@ -0,0 +1,19 @@
+syntax = "proto3";
+
+package envoy.type;
+
+import "validate/validate.proto";
+import "gogoproto/gogo.proto";
+
+option (gogoproto.equal_all) = true;
+
+// [#protodoc-title: Substitution]
+
+// Identifies a substitution.
+message Substitution {
```

Comment:
This proto is beyond the sni_cluster filter scope and we may need this to be more generic, perhaps leveraging `StringMatcher`?

cc @envoyproxy/maintainers, also cc @bmetzdorf for some thoughts from find_and_replace filter.

Reply:
Let me know if I should create a separate PR with this functionality.

Reply:
@lizan Thanks for the pointer to `StringMatcher`. I didn't know about it. It might make sense to use it in `find_and_replace`. And maybe it would also make sense to add some more features into `find_and_replace` or (more likely) a similar network filter, which enable you to extract information and augment/change envoy internal state (like metadata or target cluster). Does that make any sense?

Reply:
@bmetzdorf I haven't looked at your find/replace filter yet, but yes I do think we should try to put as much of your matching logic in common matchers as possible, since I'm gong to also be doing similar things for tap/dump (https://github.com/envoyproxy/envoy/issues/1413).